File: SkiaSharpUnoApp/SkiaSharpFunctions.cs
using Newtonsoft.Json;
using System;
using System.Threading;
using Uno.Foundation;
using Uno.Foundation.Interop;

namespace SkiaSharpUnoApp
{
	public static class SkiaSharpFunctions
	{
		private static readonly SkiaSharpFunctionsJsInterop jsInterop;

		static SkiaSharpFunctions()
		{
			jsInterop = new SkiaSharpFunctionsJsInterop();
			jsInterop.SetCurrent();
		}

		public static event EventHandler<DataReceivedEventArgs> DataReceived
		{
			add => jsInterop.DataReceived += value;
			remove => jsInterop.DataReceived -= value;
		}

		// JavaScript interop

		private class SkiaSharpFunctionsJsInterop : IJSObject, IJSObjectMetadata
		{
			private static long handleCounter = 0L;

			private readonly long jsHandle;

			public SkiaSharpFunctionsJsInterop()
			{
				jsHandle = Interlocked.Increment(ref handleCounter);
				Handle = JSObjectHandle.Create(this, this);
			}

			public JSObjectHandle Handle { get; }

			public event EventHandler<DataReceivedEventArgs> DataReceived;

			public void SetCurrent() =>
				WebAssemblyRuntime.InvokeJSWithInterop($"{this}.setCurrent();");

			private void OnDataReceived(DataReceivedEventArgs e) =>
				DataReceived?.Invoke(null, e);

			// IJSObjectMetadata

			long IJSObjectMetadata.CreateNativeInstance(IntPtr managedHandle)
			{
				WebAssemblyRuntime.InvokeJS($"SkiaSharpUnoApp.SkiaSharpFunctions.createInstance('{managedHandle}', '{jsHandle}')");
				return jsHandle;
			}

			string IJSObjectMetadata.GetNativeInstance(IntPtr managedHandle, long jsHandle) =>
				$"SkiaSharpUnoApp.SkiaSharpFunctions.getInstance('{jsHandle}')";

			void IJSObjectMetadata.DestroyNativeInstance(IntPtr managedHandle, long jsHandle) =>
				WebAssemblyRuntime.InvokeJS($"SkiaSharpUnoApp.SkiaSharpFunctions.destroyInstance('{jsHandle}')");

			object IJSObjectMetadata.InvokeManaged(object instance, string method, string parameters)
			{
				switch (method)
				{
					case nameof(OnDataReceived):
						OnDataReceived(new DataReceivedEventArgs(parameters));
						break;

					default:
						throw new ArgumentException($"Unable to execute method: {method}", nameof(method));
				}

				return null;
			}
		}
	}

	public class DataReceivedEventArgs : EventArgs
	{
		public DataReceivedEventArgs(string json)
		{
			Json = json;
		}

		public string Json { get; }

		public T Get<T>() =>
			JsonConvert.DeserializeObject<T>(Json);
	}
}

File: SkiaSharpUnoApp/WasmScripts/SkiaSharpFunctions.js
var SkiaSharpUnoApp;
(function (SkiaSharpUnoApp) {
    class SkiaSharpFunctions {
        static activeInstances = {};
        static current = undefined;

        constructor(managedHandle) {
            this.managedHandle = managedHandle;

            this.currentFragment = undefined;
            this.subscribed = false;
        }

        // JSObject
        static createInstance(managedHandle, jsHandle) {
            SkiaSharpFunctions.activeInstances[jsHandle] = new SkiaSharpFunctions(managedHandle);
        }
        static getInstance(jsHandle) {
            return SkiaSharpFunctions.activeInstances[jsHandle];
        }
        static destroyInstance(jsHandle) {
            delete SkiaSharpFunctions.activeInstances[jsHandle];
        }

        //

        setCurrent() {
            SkiaSharpFunctions.current = this;
        }

        sendData(data) {
            data = JSON.stringify(data);
            Uno.Foundation.Interop.ManagedObject.dispatch(this.managedHandle, 'OnDataReceived', data);
        }
    }

    SkiaSharpUnoApp.SkiaSharpFunctions = SkiaSharpFunctions;

})(SkiaSharpUnoApp || (SkiaSharpUnoApp = {}));
